File: leetcode/graph/1168_Optimize_Water_Distribution_in_a_Village/codes/kruskal.py
from typing import List

class Solution:
    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        parent = list(range(n + 1))

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            px, py = find(x), find(y)
            if px == py:
                return False
            parent[px] = py
            return True

        edges = [[0, i + 1, wells[i]] for i in range(n)] + pipes
        edges.sort(key=lambda x: x[2])

        cost = 0
        for u, v, w in edges:
            if union(u, v):
                cost += w
        return cost

File: leetcode/graph/1168_Optimize_Water_Distribution_in_a_Village/codes/kruskal.c
#include <stdio.h>
#include <stdlib.h>

#define MAXN 10010

int parent[MAXN];

int find(int x) {
    if (parent[x] != x)
        parent[x] = find(parent[x]);
    return parent[x];
}

int unionSet(int x, int y) {
    int px = find(x), py = find(y);
    if (px == py) return 0;
    parent[px] = py;
    return 1;
}

typedef struct {
    int u, v, cost;
} Edge;

int cmp(const void* a, const void* b) {
    return ((Edge*)a)->cost - ((Edge*)b)->cost;
}

int minCostToSupplyWater(int n, int* wells, int wellsSize, int** pipes, int pipesSize, int* pipesColSize) {
    Edge* edges = malloc(sizeof(Edge) * (pipesSize + wellsSize));
    int edgeCount = 0;

    for (int i = 0; i < n; i++) {
        edges[edgeCount++] = (Edge){0, i + 1, wells[i]};
    }

    for (int i = 0; i < pipesSize; i++) {
        edges[edgeCount++] = (Edge){pipes[i][0], pipes[i][1], pipes[i][2]};
    }

    for (int i = 0; i <= n; i++) parent[i] = i;

    qsort(edges, edgeCount, sizeof(Edge), cmp);

    int cost = 0;
    for (int i = 0; i < edgeCount; i++) {
        if (unionSet(edges[i].u, edges[i].v)) {
            cost += edges[i].cost;
        }
    }

    free(edges);
    return cost;
}
